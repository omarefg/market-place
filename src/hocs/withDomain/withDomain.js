import React from 'react';

export function withDomainBuilder(params) {
  const {
    utils: {
      formatters: { nameFunction },
    },
    hocName,
    context,
    domainAdapterPropName,
    Provider,
    useCases,
    handlersBuilder,
    useContext,
    components,
    providerMetadata
  } = params;

  return function withDomain(Component) {
    return nameFunction(hocName, (props) => {
      const childProps = {
        ...props,
        [domainAdapterPropName]: () => useContext(context),
      };

      const finalProviderProps = {
        providerMetadata,
        useCases,
        handlersBuilder,
        components,
        higherProps: props,
      };

      return (
        <Provider {...finalProviderProps}>
          <Component {...childProps} />
        </Provider>
      );
    });
  };
}
