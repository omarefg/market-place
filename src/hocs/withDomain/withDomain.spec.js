import { render } from "@testing-library/react";
import React from "react";
import { withDomainBuilder } from "./withDomain";

const ProviderMock = ({ children }) => <div>{children}</div>;
const ComponentMock = () => <div />;

const defaultParams = {
  utils: {
    formatters: {
      nameFunction: (name, body) => {
        return {
          [name](...args) {
            return body(...args);
          },
        }[name];
      },
    },
  },
  hocName: "hocName",
  context: {},
  domainAdapterPropName: "domainAdapterPropName",
  useContext: jest.fn(() => {}),
  components: {},
  Provider: ProviderMock,
  providerMetadata: {}
};

describe("@HOCS/withDomain", () => {
  it('Given valid params it must render without crashing', () => {
     // Arrange
     const params = {...defaultParams}
     // Act
     const WithDomainComponent = withDomainBuilder(params)(ComponentMock);
     const { container } = render(<WithDomainComponent/>)
     // Assert
     expect(container).toMatchSnapshot()
  })

  it("Given valid params it must return a function", () => {
    //Arrange
    const params = { ...defaultParams };
    //Act
    const withDomain = withDomainBuilder(params);
    //Assert
    expect(typeof withDomain).toBe("function");
  });

  it("Given valid params it must return a component with the name passed in `hocName`", () => {
    //Arrange
    const params = { ...defaultParams };
    //Act
    const WithDomainComponent = withDomainBuilder(params)(ComponentMock);
    // Assert
     expect(WithDomainComponent.name).toBe(defaultParams.hocName);
  });
});
