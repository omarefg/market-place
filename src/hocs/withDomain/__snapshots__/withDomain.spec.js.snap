// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@HOCS/withDomain Given valid params it must render without crashing 1`] = `
<div>
  <div>
    <div />
  </div>
</div>
`;
